kotlin-codegen/compiler/src/main/java/com/squareup/moshi/JsonClassCodeGenProcessor.kt

```diff
@@ -112,255 +84,44 @@ class JsonClassCodeGenProcessor : KotlinAbstractProcessor(), KotlinMetadataUtils
-      val getterElement = getters[name]
-      val element = fieldElement ?: setterElement ?: getterElement!!
-
-      val parameter = parameters[name]
-      var parameterIndex: Int = -1
-      var parameterElement: VariableElement? = null
-      if (parameter != null) {
-        parameterIndex = protoConstructor.valueParameterList.indexOf(parameter)
-        parameterElement = constructor.parameters[parameterIndex]
-      }
-
-      val annotationHolder = annotationHolders[property]
-
-      if (property.visibility != Visibility.INTERNAL
-          && property.visibility != Visibility.PROTECTED
-          && property.visibility != Visibility.PUBLIC) {
-        messager.printMessage(ERROR, "property $name is not visible", element)
-        return null
-      }
-
-      val hasDefault = parameter?.declaresDefaultValue ?: true
-
-      if (Modifier.TRANSIENT in element.modifiers) {
-        if (!hasDefault) {
-          messager.printMessage(
-              ERROR, "No default value for transient property $name", element)
-          return null
-        }
-        continue // This property is transient and has a default value. Ignore it.
-      }
-
-      if (!property.hasSetter && parameter == null) {
-        continue // This property is not settable. Ignore it.
+    val properties = mutableMapOf<String, PropertyGenerator>()
+    for (property in type.properties.values) {
+      val generator = property.generator(messager)
+      if (generator != null) {
+        properties[property.name] = generator
       }
-
-      val delegateKey = DelegateKey(
-          property.returnType.asTypeName(nameResolver, classProto::getTypeParameter, true),
-          jsonQualifiers(element, annotationHolder, parameterElement))
-
-      propertiesByName[name] = PropertyGenerator(
-          delegateKey,
-          name,
-          jsonName(name, element, annotationHolder, parameterElement),
-          parameterIndex,
-          hasDefault,
-          property.returnType.asTypeName(nameResolver, classProto::getTypeParameter))
     }
 
-    for (parameterElement in constructor.parameters) {
-      val name = parameterElement.name
-      val valueParameter = parameters[name]!!
-      if (properties[name] == null && !valueParameter.declaresDefaultValue) {
+    for ((name, parameter) in type.constructor.parameters) {
+      if (type.properties[parameter.name] == null && !parameter.proto.declaresDefaultValue) {
         messager.printMessage(
-            ERROR, "No property for required constructor parameter $name", parameterElement)
+            ERROR, "No property for required constructor parameter $name", parameter.element)
         return null
       }
     }
 
     // Sort properties so that those with constructor parameters come first.
-    val propertyGenerators = propertiesByName.values.toMutableList()
-    propertyGenerators.sortBy {
+    val sortedProperties = properties.values.toMutableList()
+    sortedProperties.sortBy {
```

Comment:
I know this is existing code, but you can skip `toMutableList()` by just calling `sortedBy { }` directly on `values`.

Reply:
Will fix in follow up!